modules/shapefile/src/lib/parsers/parse-shapefile.js

```diff
@@ -5,26 +5,44 @@ import {zipBatchIterators} from '../streaming/zip-batch-iterators';
 import {SHPLoader} from '../../shp-loader';
 import {DBFLoader} from '../../dbf-loader';
 
+// eslint-disable-next-line max-statements
 export async function* parseShapefileInBatches(asyncIterator, options, context) {
+  const {metadata} = options;
   const {reproject = false, _targetCrs = 'WGS84'} = (options && options.gis) || {};
   const {parseInBatches, fetch, url} = context;
   const {shx, cpg, prj} = await loadShapefileSidecarFiles(options, context);
 
   // parse geometries
-  const shapeIterator = await parseInBatches(asyncIterator, SHPLoader, options); // {shp: shx}
-  const shapeHeader = (await shapeIterator.next()).value;
+  const shapeIterator = await parseInBatches(asyncIterator, SHPLoader, options);
 
   // parse properties
   let propertyIterator;
-  let iterator;
   const dbfResponse = await fetch(replaceExtension(url, 'dbf'));
   if (dbfResponse.ok) {
     propertyIterator = await parseInBatches(dbfResponse, DBFLoader, {
+      ...options,
       dbf: {encoding: cpg || 'latin1'}
     });
+  }
+
+  // When `options.metadata` is `true`, there's an extra initial `metadata`
+  // object before the iterator starts. zipBatchIterators expects to receive
+  // batches of Array objects, and will fail with non-iterable batches, so it's
+  // important to skip over the first batch.
+  if (metadata) {
+    await shapeIterator.next();
+    if (propertyIterator) {
```

Comment:
Isn't there a `type: 'metadata'` field on the metadata batch - I feel that would be a much more robust way to filter out metadata batches than just assuming they are there.

Reply:
I modified `parseDBFInBatches` so that it also emits an initial metadata batch like the `SHPLoader`. Now I can request a first batch from each of SHP/DBF, and if that first batch contains `batchType: 'metadata'`, then I know the header will be the _next_ batch. Adding the DBF header batch allows me to "peek" at the next batch without looking at actual records that I'd then want to put back on the stack before passing to `zipBatchIterators`.